openquake/hazardlib/shakemap/gmfs.py

```diff
@@ -151,41 +152,131 @@ def cholesky(spatial_cov, cross_corr):
     return numpy.linalg.cholesky(numpy.array(LLT))
 
 
-def to_gmfs(shakemap, spatialcorr, crosscorr, site_effects, trunclevel,
-            num_gmfs, seed, imts=None):
+gmfs_calculation_methods = CallableDict()
+
+MAXSITES = 1000
+CORRELATION_MATRIX_TOO_LARGE = '''\
+You have a correlation matrix which is too large: %s > %d.
+To avoid that, set a proper `region_grid_spacing` so that your exposure
+involves less sites.'''
+
+
+@gmfs_calculation_methods.add('Silva&Horspool')
+def gmfs_calculation_methods_SH(kind, shakemap, imts, spatialcorr,
+                                crosscorr, cholesky_limit):
     """
-    :returns: (IMT-strings, array of GMFs of shape (R, N, E, M)
+    Implementation of paper by Silva and Horspool 2019
+    https://onlinelibrary.wiley.com/doi/abs/10.1002/eqe.3154?af=R
+
+    :param shakemap: site coordinates with shakemap values
+    :param imts: list of required imts
+    :param spatialcorr: 'no', 'yes' or 'full'
+    :param crosscorr: 'no', 'yes' or 'full'
+    :returns: F(Z, mu) to calculate gmfs
     """
-    N = len(shakemap)  # number of sites
-    std = shakemap['std']
-    if imts is None or len(imts) == 0:
-        imts = std.dtype.names
-    else:
-        imts = [imt for imt in imts if imt in std.dtype.names]
-    val = {imt: numpy.log(shakemap['val'][imt]) for imt in imts}
-    imts_ = [imt.from_string(name) for name in imts]
-    M = len(imts_)
-    cross_corr = cross_correlation_matrix(imts_, crosscorr)
-    mu = numpy.array([numpy.ones(num_gmfs) * val[str(imt)][j]
-                      for imt in imts_ for j in range(N)])
+    # checks
+    N = len(shakemap)
+    M = len(imts)
+    if spatialcorr != 'no' and N > MAXSITES:
+        # hard-coded, heuristic
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE %
+                         (N, MAXSITES))
+    if N * M > cholesky_limit:
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE % (
+            '%d x %d' % (M, N), cholesky_limit))
+
+    # Cross Correlation
+    cross_corr = cross_correlation_matrix(imts, crosscorr)
+
+    # Spatial Correlation and Covariance
     dmatrix = geo.geodetic.distance_matrix(shakemap['lon'], shakemap['lat'])
-    spatial_corr = spatial_correlation_array(dmatrix, imts_, spatialcorr)
-    stddev = [std[str(imt)] for imt in imts_]
-    for im, std in zip(imts_, stddev):
+    spatial_corr = spatial_correlation_array(dmatrix, imts, spatialcorr)
+
+    stddev = [shakemap['std'][str(imt)] for imt in imts]
+    for im, std in zip(imts, stddev):
         if std.sum() == 0:
             raise ValueError('Cannot decompose the spatial covariance '
                              'because stddev==0 for IMT=%s' % im)
     spatial_cov = spatial_covariance_array(stddev, spatial_corr)
+
+    # Cholesky Decomposition
     L = cholesky(spatial_cov, cross_corr)  # shape (M * N, M * N)
+
+    # mu has unit (pctg), L has unit ln(pctg)
+    def calculate_gmfs(Z, mu):
+        mu = numpy.log(mu)
+        return numpy.exp(L @ Z + mu) / PCTG
+
```

Comment:
every call to `gmfs_calculation_methods` returns a function (`calculate_gmfs`). In these two cases the function looks the same (except that one has L@Z+mu and the other one sig@Z+mu). But I will add for example something along those lines:

`@gmfs_calculation_methods.add('MMI')
def gmfs_calculation_methods_basic(kind, shakemap, imts):
    # create diag matrix with std values
    std = numpy.array(shakemap['std']['MMI'])
    sig = numpy.diag(std.flatten())  # shape (M*N, M*N)

    # mu has unit (pctg), sig has unit ln(pctg)
    def calculate_gmfs(Z, mu):
        return sig @ Z + mu

    return calculate_gmfs`

Reply:
I do not like that `gmfs_calculation_methods_basic` returns a function. It should just return the result. If you need the function you can just call `gmfs_calculation_methods_basic[kind]`. BTW, I do not like the name `gmfs_calculation_methods_basic`. It should be something like `shakemap_to_gmfs`.

Reply:
I don't think this helps readability and gmf_dict is huge by now since I also have to put in Z and mu. 

But you're the boss, I made the changes as requested.